Add gatk somatic paired calling subworkflow

GCJMackenzie opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

Closes #1047 

This PR adds a subworkflow for performing somatic variant calling using gatk4 mutect in paired tumor/normal mode. The f1r2 output of mutect2 is used with learnreadorientationmodel to generate the artifact priors. The input files are run through getpileupsummaries and then calculatecontamination to produce the contamination and segmentation tables. These tables are passed along with the mutect2 calls and artifact priors into filtermutectcalls for filtering.

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

maxulysse commented on subworkflows/nf-core/gatk_paired_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,110 @@
+
+include { GATK4_MUTECT2                   as MUTECT2 }                   from '../../../modules/gatk4/mutect2/main'                   addParams( options: params.mutect2_options )
+include { GATK4_LEARNREADORIENTATIONMODEL as LEARNREADORIENTATIONMODEL } from '../../../modules/gatk4/learnreadorientationmodel/main' addParams( options: params.learnorientation_options )
+include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES }        from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileup_options )
+include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES_NORMAL}  from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileupnorm_options )
```
I'd specify the `GETPILEUPSUMMARIES` as `GETPILEUPSUMMARIES_TUMOR` as you did for the `GETPILEUPSUMMARIES_NORMAL`

maxulysse commented:
I think I use the `tumor` spelling over the `tumour` one.
But apart from that, it's looking good.

I think in Sarek in my mapping subworkflow, I started specifying only the `tumor` processes and keep the `normal` ones as default.
It could be good to have a proper discussion on how to tackle that in the future, just to be sure that we keep the same logic everywhere.

Should we specify only the `tumor` processes, only the `normal` one or both
@FriederikeHanssen what is your take for that?

FriederikeHanssen commented on subworkflows/nf-core/gatk_paired_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,110 @@
+
+params.mutect2_options          = [:]
+params.learnorientation_options = [:]
+params.getpileup_options        = [suffix: '_tumour']
```
I think in other processes we went along with `T` and `N`, didn't we @maxulysse ?

FriederikeHanssen replied:
or rather nothing and somatic ( in the case of somatic calling of course)

FriederikeHanssen commented:
I agree on the spelling, just to keep it consistent in the pipeline.

Hm I kind of prefer being explicit in naming. So we could have `tumor`, `normal` and then either stick with `tumor_vs_normal` or go straight with somatic? For the output I think we should stick with the sample IDs. I like using it as sanity check when investigating the results.

FriederikeHanssen commented:
@chelauk I think you were involved at some point in the Mutect2 usage in the previous sarek version. Do you have any thoughts on the subworkflow?

GCJMackenzie commented:
I prefer explicitly using tumor and normal as well. I think tumor_vs_normal would be better than just saying somatic, as I have a tumor-only version of this subworkflow, which is also somatic, so tumor_vs_normal would help keep the two distinct.

chelauk commented:
> @chelauk I think you were involved at some point in the Mutect2 usage in the previous sarek version. Do you have any thoughts on the subworkflow?

looks amazing, I can't wait to try it.  I am most comfortable with 'tumor' 'normal' annotation

GCJMackenzie commented:
I have changed 'tumour' to 'tumor' in this PR and the tumor_only PR.
I have also renamed the GETPILEUPSUMMARIES to GETPILEUPSUMMARIES_NORMAL
Finally I renamed the subworkflow from 'gatk_paired_somatic_variant_calling' to 'gatk_tumor_normal_somatic_variant_calling'

GCJMackenzie commented:
@FriederikeHanssen on the note of using the sample IDs for the ouputs, wouldn't they be passed in as the meta:ID anyway? So when running a normal run we would have {sample_ID}_tumor.pileups.table as things stand currently. 

If you mean the individual filenames for the tumor and normal recalibrated bams, could we pass in the file_name prefixes to the subworkflow via optional arguments? i.e. changing the current prefixes for the GETPILEUPSUMMARIES modules from '_tumor' to '${sampleID}_tumor' or something when the subworkflow is called by the main workflow?

FriederikeHanssen commented:
🤦‍♀️ I was thinking about how we do the publish dir output naming in sarek. Nevermind me, I was apparently somewhere else 😆

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+params.mutect2_options          = [:]
+params.learnorientation_options = [:]
+params.getpileup_options        = [suffix: '_tumor']
+params.getpileupnorm_options    = [suffix: '_normal']
```
```suggestion
params.getpileup_normal_options    = [suffix: '_normal']
```

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+
+include { GATK4_MUTECT2                   as MUTECT2 }                   from '../../../modules/gatk4/mutect2/main'                   addParams( options: params.mutect2_options )
+include { GATK4_LEARNREADORIENTATIONMODEL as LEARNREADORIENTATIONMODEL } from '../../../modules/gatk4/learnreadorientationmodel/main' addParams( options: params.learnorientation_options )
+include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES_TUMOR }  from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileup_options )
```
```suggestion
include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES_TUMOR }  from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileup_tumor_options )
```

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+include { GATK4_MUTECT2                   as MUTECT2 }                   from '../../../modules/gatk4/mutect2/main'                   addParams( options: params.mutect2_options )
+include { GATK4_LEARNREADORIENTATIONMODEL as LEARNREADORIENTATIONMODEL } from '../../../modules/gatk4/learnreadorientationmodel/main' addParams( options: params.learnorientation_options )
+include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES_TUMOR }  from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileup_options )
+include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES_NORMAL}  from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileupnorm_options )
```
```suggestion
include { GATK4_GETPILEUPSUMMARIES        as GETPILEUPSUMMARIES_NORMAL}  from '../../../modules/gatk4/getpileupsummaries/main'        addParams( options: params.getpileup_normal_options )
```

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/nextflow.config:
```diff
@@ -0,0 +1,6 @@
+params.mutect2_options           = [:]
+params.learnorientation_options  = [:]
+params.getpileup_options         = [:]
+params.getpileupnorm_options     = [:]
```
```suggestion
params.getpileup_tumor_options         = [:]
params.getpileup_normal_options     = [:]
```

maxulysse commented:
I suggested more "improvements" (and I emphasize the double quotes, as I'm really not sure it's really improvements).
But basically, I think it'll make more sense to have `_normal` and `_tumor` in the `params.*_options` for each module too.

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+
+params.mutect2_options          = [:]
+params.learnorientation_options = [:]
+params.getpileup_tumor_options        = [suffix: '_tumor']
+params.getpileup_normal_options    = [suffix: '_normal']
```
```suggestion
params.getpileup_tumor_options  = [suffix: '_tumor']
params.getpileup_normal_options = [suffix: '_normal']
```

maxulysse replied:
sorry, I messed up the indents

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/meta.yml:
```diff
@@ -0,0 +1,127 @@
+      type: file
+      description: Population vcf of germline sequencing, containing allele fractions.
+      pattern: "*.vcf.gz"
+  - germline_resource_idx:
```
@FriederikeHanssen I think I usually go for `_tbi` for the tabix index.
What's your take on that?

FriederikeHanssen replied:
yes lets stick with tbi, think this is what we used in strelka etc as well

GCJMackenzie replied:
Will need to change this in a few modules I think, will try remember to do that when we update the test data. Just to clarify, it should be "germline_resource_tbi"? Assuming it should be the same for "fastaidx" and "panel_of_normals_idx" as well?

maxulysse replied:
yes, `panel_of_normals_idx` -> `panel_of_normals_tbi` or even `pon_tbi` actually.
and `fastaidx` -> `fai`

FriederikeHanssen replied:
I like panel of normal. We have complete names for the other files as well

GCJMackenzie commented:
variables for the index files should be updated across this PR now

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+
+workflow GATK_TUMOR_NORMAL_SOMATIC_VARIANT_CALLING {
+    take:
+    ch_mutect2_in             // channel: [ val(meta), [ input ], [ input_index ], [which_norm] ]
+    fasta                     // channel: /path/to/reference/fasta
+    fai                       // channel: /path/to/reference/fasta/index
+    dict                      // channel: /path/to/reference/fasta/dictionary
+    germline_resource         // channel: /path/to/germline/resource
+    germline_resource_tbi     // channel: /path/to/germline/index
+    panel_of_normals          // channel: /path/to/panel/of/normals
+    panel_of_normals_tbi      // channel: /path/to/panel/of/normals/index
+    interval_file             // channel: /path/to/interval/file
+
+
+    main:
+    ch_versions = Channel.empty()
+    input = channel.from(ch_mutect2_in)
```
```suggestion
    input                     // channel: [ val(meta), [ input ], [ input_index ], [which_norm] ]
    fasta                     // channel: /path/to/reference/fasta
    fai                       // channel: /path/to/reference/fasta/index
    dict                      // channel: /path/to/reference/fasta/dictionary
    germline_resource         // channel: /path/to/germline/resource
    germline_resource_tbi     // channel: /path/to/germline/index
    panel_of_normals          // channel: /path/to/panel/of/normals
    panel_of_normals_tbi      // channel: /path/to/panel/of/normals/index
    interval_file             // channel: /path/to/interval/file


    main:
    ch_versions = Channel.empty()
```

maxulysse replied:
I'd even go for that

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+    //Generate pileup summary tables using getepileupsummaries. tumor sample should always be passed in as the first input and input list entries of ch_mutect2_in,
+    //to ensure correct file order for calculatecontamination.
+    //
+    pileup_tumor_input = channel.from(ch_mutect2_in)
+    pileup_tumor_input = pileup_tumor_input.map {
+        meta, input_file, input_index, which_norm ->
+        [meta, input_file[0], input_index[0]]
+    }
+
+    pileup_normal_input = channel.from(ch_mutect2_in)
+    pileup_normal_input = pileup_normal_input.map {
+        meta, input_file, input_index, which_norm ->
+        [meta, input_file[1], input_index[1]]
+    }
```
```suggestion
    pileup_tumor_input = input.map {
        meta, input_file, input_index, which_norm ->
        [meta, input_file[0], input_index[0]]
    }

    pileup_normal_input = input.map {
        meta, input_file, input_index, which_norm ->
        [meta, input_file[1], input_index[1]]
    }
```

maxulysse replied:
Shouldn't that be simpler too?

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+    //
+    //Contamination and segmentation tables created using calculatecontamination on the pileup summary table.
+    //
+    ch_pileup     = GETPILEUPSUMMARIES_TUMOR.out.table.collect()
+    ch_pileup2    = GETPILEUPSUMMARIES_NORMAL.out.table.collect()
```
instead of `_pileup` and `_pileup2` I'd go for `_pileup_tumor` and `_pileup_normal`

GCJMackenzie commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,111 @@
+    //
+    //Contamination and segmentation tables created using calculatecontamination on the pileup summary table.
+    //
+    ch_pileup     = GETPILEUPSUMMARIES_TUMOR.out.table.collect()
+    ch_pileup2    = GETPILEUPSUMMARIES_NORMAL.out.table.collect()
+    ch_calccon_in = ch_pileup.combine(ch_pileup2, by: 0)
```
```suggestion
    ch_pileup_tumor     = GETPILEUPSUMMARIES_TUMOR.out.table.collect()
    ch_pileup_normal    = GETPILEUPSUMMARIES_NORMAL.out.table.collect()
    ch_calccon_in = ch_pileup_tumor.combine(ch_pileup_normal, by: 0)
```

GCJMackenzie commented on tests/subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,25 @@
+include { GATK_TUMOR_NORMAL_SOMATIC_VARIANT_CALLING } from '../../../../subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main' addParams( [:]  )
+
+workflow test_gatk_tumor_normal_somatic_variant_calling {
+    ch_mutect2_in = [
```
```suggestion
    input = [
```

GCJMackenzie commented on tests/subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,25 @@
+
+include { GATK_TUMOR_NORMAL_SOMATIC_VARIANT_CALLING } from '../../../../subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main' addParams( [:]  )
+
+workflow test_gatk_tumor_normal_somatic_variant_calling {
+    ch_mutect2_in = [
+                    [ [ id:'test'], // meta map
+                    [ file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_recalibrated_sorted_bam'], checkIfExists: true) , file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam'], checkIfExists: true)],
+                    [ file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true) , file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true)],
+                    ["testN"]
+                    ]
+                    ]
```
```suggestion
workflow test_gatk_tumor_normal_somatic_variant_calling {
    input =         [
                    [ [ id:'test'], // meta map
                    [ file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_recalibrated_sorted_bam'], checkIfExists: true) , file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam'], checkIfExists: true)],
                    [ file(params.test_data['homo_sapiens']['illumina']['test2_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true) , file(params.test_data['homo_sapiens']['illumina']['test_paired_end_recalibrated_sorted_bam_bai'], checkIfExists: true)],
                    ["testN"]
                    ]
                    ]
```

GCJMackenzie commented on tests/subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,25 @@
+    panel_of_normals_tbi  = file(params.test_data['homo_sapiens']['genome']['mills_and_1000g_indels_vcf_gz_tbi'],       checkIfExists: true)
+    interval_file         = file(params.test_data['homo_sapiens']['genome']['genome_interval_list'],                    checkIfExists: true)
+
+    GATK_TUMOR_NORMAL_SOMATIC_VARIANT_CALLING ( ch_mutect2_in, fasta, fai, dict, germline_resource, germline_resource_tbi, panel_of_normals, panel_of_normals_tbi, interval_file )
```
```suggestion
    GATK_TUMOR_NORMAL_SOMATIC_VARIANT_CALLING ( input, fasta, fai, dict, germline_resource, germline_resource_tbi, panel_of_normals, panel_of_normals_tbi, interval_file )
```

GCJMackenzie commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,108 @@
+    //
+    //Perform variant calling using mutect2 module in tumor single mode.
+    //
+    MUTECT2 ( input, false, false, false, [], fasta, fai, dict, germline_resource, germline_resource_tbi, panel_of_normals, panel_of_normals_tbi )
```
```suggestion
    mutect2_input = channel_from(input)
    MUTECT2 ( mutect2_input, false, false, false, [], fasta, fai, dict, germline_resource, germline_resource_tbi, panel_of_normals, panel_of_normals_tbi )
```

GCJMackenzie commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/main.nf:
```diff
@@ -0,0 +1,108 @@
+    //Generate pileup summary tables using getepileupsummaries. tumor sample should always be passed in as the first input and input list entries of ch_mutect2_in,
+    //to ensure correct file order for calculatecontamination.
+    //
+    pileup_tumor_input = input.map {
+        meta, input_file, input_index, which_norm ->
+        [meta, input_file[0], input_index[0]]
+    }
+
+    pileup_normal_input = input.map {
+        meta, input_file, input_index, which_norm ->
+        [meta, input_file[1], input_index[1]]
```
```suggestion
    pileup_tumor_input = channel_from(input).map {
        meta, input_file, input_index, which_norm ->
        [meta, input_file[0], input_index[0]]
    }

    pileup_normal_input = channel_from(input).map {
        meta, input_file, input_index, which_norm ->
        [meta, input_file[1], input_index[1]]
```

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/meta.yml:
```diff
@@ -0,0 +1,127 @@
+  - fai:
+      type: file
+      description: Index of reference fasta file
+      pattern: "fasta.fai"
```
```suggestion
      pattern: "*.fasta.fai"
```

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/meta.yml:
```diff
@@ -0,0 +1,127 @@
+  - germline_resource_tbi:
+      type: file
+      description: Index file for the germline resource.
+      pattern: "*.vcf.gz_tbi"
```
```suggestion
      pattern: "*.vcf.gz.tbi"
```

maxulysse commented on subworkflows/nf-core/gatk_tumor_normal_somatic_variant_calling/meta.yml:
```diff
@@ -0,0 +1,127 @@
+  - panel_of_normals_tbi:
+      type: file
+      description: Index for the panel of normals.
+      pattern: "*.vcf.gz_tbi"
```
```suggestion
      pattern: "*.vcf.gz.tbi"
```

maxulysse commented:
same comments than https://github.com/nf-core/modules/pull/1064

GCJMackenzie commented:
Thanks @maxulysse will update the subworkflows.yml after lunch.
